File: bms/go/geo/geo_r2_Rect_ApproxEqual/geo_r2_Rect_ApproxEqual.go
package test

import (
	"math"
)

var ()

const (
	geo_s2_epsilon = 1e-15
	geo_r1_epsilon = 1e-15
)

func (r geo_r2_Rect) ApproxEqual(r2 geo_r2_Rect) bool {
	return r.X.ApproxEqual(r2.X) && r.Y.ApproxEqual(r2.Y)
}

func (i geo_r1_Interval) ApproxEqual(other geo_r1_Interval) bool {
	if i.IsEmpty() {
		return other.Length() <= 2*geo_r1_epsilon
	}
	if other.IsEmpty() {
		return i.Length() <= 2*geo_r1_epsilon
	}
	return math.Abs(other.Lo-i.Lo) <= geo_r1_epsilon &&
		math.Abs(other.Hi-i.Hi) <= geo_r1_epsilon
}

func (i geo_r1_Interval) IsEmpty() bool { return i.Lo > i.Hi }

func (i geo_r1_Interval) Length() float64 { return i.Hi - i.Lo }

type geo_r2_Rect struct {
	X, Y geo_r1_Interval
}

type geo_r1_Interval struct {
	Lo, Hi float64
}

File: translations_only/fail/go/geo/geo_r2_Rect_ApproxEqual/mistral/BaseRepair/geo_r2_Rect_ApproxEqual.rs

pub struct GeoR2Rect {
    x: GeoR1Interval,
    y: GeoR1Interval,
}

pub struct GeoR1Interval {
    lo: f64,
    hi: f64,
}

const GEO_S2_EPSILON: f64 = 1e-15;
const GEO_R1_EPSILON: f64 = 1e-15;

impl GeoR1Interval {
    pub fn approx_equal(&self, other: &GeoR1Interval) -> bool {
        if self.is_empty() {
            return other.length() <= 2. * GEO_R1_EPSILON;
        }
        if other.is_empty() {
            return self.length() <= 2. * GEO_R1_EPSILON;
        }
        f64::abs(other.lo - self.lo) <= GEO_R1_EPSILON && f64::abs(other.hi - self.hi) <= GEO_R1_EPSILON
    }

    pub fn is_empty(&self) -> bool {
        self.lo > self.hi
    }

    pub fn length(&self) -> f64 {
        self.hi - self.lo
    }
}

impl GeoR2Rect {
    pub fn approx_equal(&self, r2: &GeoR2Rect) -> bool {
        self.x.approx_equal(&r2.x) && self.y.approx_equal(&r2.y)
    }
}
